src/platforms/esp32/components/avm_builtins/i2c_resource.c

```diff
@@ -0,0 +1,743 @@
+    conf.sda_io_num = term_to_int32(sda);
+    conf.sda_pullup_en = GPIO_PULLUP_ENABLE;
+    conf.scl_pullup_en = GPIO_PULLUP_ENABLE;
+    conf.master.clk_speed = term_to_int32(clock_speed_hz);
+
+    esp_err_t err;
+    err = i2c_param_config(i2c_num, &conf);
+    if (UNLIKELY(err != ESP_OK)) {
+        ESP_LOGE(TAG, "Failed to initialize I2C parameters.  err=%i", err);
+        if (UNLIKELY(memory_ensure_free(ctx, TUPLE_SIZE(2)) != MEMORY_GC_OK)) {
+            ESP_LOGW(TAG, "Failed to allocate memory: %s:%i.\n", __FILE__, __LINE__);
+            return OUT_OF_MEMORY_ATOM;
+        }
+        // TODO breaks i2c::open/1 API
+        return create_error_tuple(ctx, term_from_int(err));
+    }
+
+    //
+    // Install the I2C driver
+    //
+
+    err = i2c_driver_install(i2c_num, I2C_MODE_MASTER, 0, 0, 0);
+    if (UNLIKELY(err != ESP_OK)) {
+        ESP_LOGE(TAG, "Failed to install I2C driver.  err=%i", err);
+        if (UNLIKELY(memory_ensure_free(ctx, TUPLE_SIZE(2)) != MEMORY_GC_OK)) {
+            ESP_LOGW(TAG, "Failed to allocate memory: %s:%i.\n", __FILE__, __LINE__);
+            return OUT_OF_MEMORY_ATOM;
+        }
+        // TODO breaks i2c::open/1 API
+        return create_error_tuple(ctx, term_from_int(err));
+    }
+
+    ESP_LOGI(TAG, "I2C driver installed using I2C port %i", i2c_num);
+
+    //
+    // allocate and initialize the Nif resource
+    //
+
+    struct I2CResource *rsrc_obj = enif_alloc_resource(i2c_resource_type, sizeof(struct I2CResource));
+    if (IS_NULL_PTR(rsrc_obj)) {
+        i2c_driver_delete(i2c_num);
+        ESP_LOGW(TAG, "Failed to allocate memory: %s:%i.\n", __FILE__, __LINE__);
+        RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+    }
+    rsrc_obj->transmitting_pid = term_invalid_term();
+    rsrc_obj->i2c_num = i2c_num;
+
+    if (UNLIKELY(memory_ensure_free(ctx, TERM_BOXED_RESOURCE_SIZE) != MEMORY_GC_OK)) {
+        i2c_driver_delete(i2c_num);
+        enif_release_resource(rsrc_obj);
+        ESP_LOGW(TAG, "Failed to allocate memory: %s:%i.", __FILE__, __LINE__);
+        RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+    }
+    term obj = enif_make_resource(erl_nif_env_from_context(ctx), rsrc_obj);
+    enif_release_resource(rsrc_obj);
+
+    //
+    // Return result
+    //
+
+    // {'$i2c', Resource :: resource(), Ref :: reference()} :: i2c()
+    size_t requested_size = TUPLE_SIZE(3) + REF_SIZE;
+    if (UNLIKELY(memory_ensure_free_with_roots(ctx, requested_size, 1, &obj, MEMORY_CAN_SHRINK) != MEMORY_GC_OK)) {
+        i2c_driver_delete(i2c_num);
+        ESP_LOGW(TAG, "Failed to allocate memory: %s:%i.", __FILE__, __LINE__);
+        RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+    }
+
+    term i2c = term_alloc_tuple(3, &ctx->heap);
+    term_put_tuple_element(i2c, 0, globalcontext_make_atom(global, I2C_ATOMSTR));
+    term_put_tuple_element(i2c, 1, obj);
+    uint64_t ref_ticks = globalcontext_get_ref_ticks(ctx->global);
+    term ref = term_from_ref_ticks(ref_ticks, &ctx->heap);
+    term_put_tuple_element(i2c, 2, ref);
```

Comment:
I'm not sure about the purpose of this ref.

Reply:
Currently, Nif resources are represented using binaries (in order to leverage reference counting).  However, due to this implementation, resource objects always test true for equality.  Therefore, a reference (or some unique instance identifier) needs to be thrown into the term representing a resource in order to guarantee uniqueness of resource instances.  This is a known "feature" of the resource-based Nif implementation.

Reply:
This was a limitation of the old OTP resource implementation that was wrongly used as a model for AtomVM. We should migrate to magic numbers and fix comparison.

Reply:
I think it will be fine here now, do we have any open issue for moving forward to magic references?